---
title: "Problem 3"
output: html_document
date: "2023-11-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(tidyverse)
```

1. Use R and ggplot2 package to draw a a plot of number of genes per chromosome in human genomes.
This task requires the data file Homo_sapiens.gene.info.gz. You need to use columns 3 and 7 indicating Symbol and chromosome respectively. 
Save the plot to PDF file.
```{r}
dat <- read.delim("../data/Homo_sapiens.gene_info.gz", header = T)

gpc <- dat |> 
  select(Symbol, chromosome) |> 
  summarise(.by = chromosome, gene_count = n_distinct(Symbol)) |> 
  mutate(chromosome = factor(chromosome, 
                             levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10",
                                        "11", "12", "13", "14", "15", "16", "17", "18", "19",
                                        "20", "21", "22", "X", "Y", "MT", "Un"))) |> 
  drop_na()

ggplot(gpc, aes(x = chromosome, y = gene_count)) +
  geom_col() +
  labs(y = "Gene Count", x = "Chromosome", title = "Number of genes in each chromosome") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5))

# ggsave("num_of_genes_per_chr.pdf", height = 3.5, width = 7, units = "in")
```

2. The longer chromosome might have higher frequency of genes most probably by chance. We will test this hypothesis. You can find the length information of each chromosome here: https://www.ncbi.nlm.nih.gov/grc/human/data. Is there any correlation between number of genes and chromosome length?
Also, plot the regression data with the trendline. Calculate the RË†2 and other statistics to determine
whether the fit is significant. From the regression equation estimate the number of genes expected for
each chromosome. Then evaluate whether any chromosome has higher or lower concentration of genes.
Hint: Ignore MT and UN. You need to find 95% confidence interval of the regression. Read more about
it here: https://rpubs.com/aaronsc32/regression-confidence-prediction-intervals
```{r}
# download.file("https://ftp.ncbi.nlm.nih.gov/pub/grc/human/GRC/GRCh38.p14_GenBank/GCA_000001405.29_GRCh38.p14_assembly_report.txt", destfile = "GRCh38.p14_assembly_report")

GRCh38 <- read_delim("GRCh38.p14_assembly_report", 
    delim = "\t", escape_double = FALSE, 
    col_names = FALSE, comment = "#", trim_ws = TRUE)

chr_lens <- GRCh38[1:24,c(3,9)]
names(chr_lens) <- c("chromosome", "length")

full_df <- left_join(chr_lens, gpc, by = "chromosome") |> 
  mutate(chromosome = factor(chromosome, 
                             levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10",
                                        "11", "12", "13", "14", "15", "16", "17", "18", "19",
                                        "20", "21", "22", "X", "Y")))

cor(x = full_df$length, y = full_df$gene_count)

chr_lm <- lm(gene_count ~ length, full_df)
summary(chr_lm)

ggplot(full_df, aes(x = length, y = gene_count)) +
  geom_smooth(method = 'lm', formula = 'y ~ x', alpha = 1/5) +
  geom_point(aes(color = chromosome), show.legend = F) +
  labs(y = "Gene Count", x = "Chromosome Length (bp)", color = "Chromosome") +
  scale_x_continuous(labels = scales::number_format(big.mark = ",")) +
  scale_y_continuous(labels = scales::number_format(big.mark = ",")) +
  geom_text(aes(label = chromosome), nudge_x = 5000000, nudge_y = 500) +
  theme_bw() +
  scale_color_viridis_d(option = "turbo")
  
preds <- cbind(full_df, predict(chr_lm, full_df, interval = "confidence", level = 0.95) |> as.data.frame()) |> 
  mutate(sig_diff = ifelse(gene_count < lwr | gene_count > upr, TRUE, FALSE))

# chromosomes with gene counts outside of the regression model's 95% confidence intervals
cat(preds$chromosome[which(preds$sig_diff == TRUE)])

# percentage of chromosomes out of confidence intervals
sum(preds$sig_diff == TRUE)/length(preds$sig_diff)*100
```

There is considerable positive correlation between `gene count` and `chromosome length` with a pearson coefficient of 0.79 and corresponding R^2 of 0.63. Linear regression suggests this correlation is statistically significant (p = 3.61e-6) and allows us to determine that chromosome length explains approximately 61.1% of the variation in gene counts, according to an adjusted R^2 statistic. Notably, 10 out of 24 chromosomes (41.7%), including chromosome 1, had gene counts that were outside of the regression model's 95% confidence intervals. **In other words: chromosome 1 appears to have the highest number of genes not just because it has the longest length.**

3. In earlier problem we calculated the frequency of disease variant in each gene in human genome. Can
you evaluate whether any human chromosome has more concentration of disease variants?
```{r}
chr_locs <- dat |> 
  select(gene_name = Symbol, chromosome)

dis_vars <- read_tsv("../data/dis_vars.tsv") |> 
  left_join(chr_locs, by = "gene_name") |> 
  mutate(chromosome = factor(chromosome, 
                             levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10",
                                        "11", "12", "13", "14", "15", "16", "17", "18", "19",
                                        "20", "21", "22", "X", "Y"))) |> 
  summarise(.by = chromosome, var_count = sum(n)) |> 
  drop_na() |> 
  left_join(full_df, by = "chromosome") |> 
  mutate(var_freq = var_count/gene_count)

ggplot(dis_vars, aes(x = chromosome, y = var_freq)) +
  geom_col(aes(fill = chromosome), show.legend = F) +
  geom_hline(yintercept = quantile(dis_vars$var_freq, 0.5, names = F)) +
  geom_hline(yintercept = quantile(dis_vars$var_freq, 0.75, names = F) + IQR(dis_vars$var_freq)*1.5, 
             linetype = "dashed") +
  geom_hline(yintercept = quantile(dis_vars$var_freq, 0.25, names = F) - IQR(dis_vars$var_freq)*1.5, 
             linetype = "dashed") +
  labs(y = "Disease Variant Frequencey\n(variant count / gene count)", x = "Chromosome",
       caption = "Median = solid black line; Upper whisker (Q3 + 1.5*IQR) = top dashed line; Lower wisker (Q1 - 1.5*IQR) = bottom dashed line") +
  theme_bw()
```

As seen in the plot, the disease variant frequency for chr X is an extreme outlier. In other words: chromosome X has significantly more recorded disease causing mutations relative to its total number of genes when compared to the other chromosomes.